File: longest_palindrome/longest_palindrome.ts
//My Solution
const longestPalindrome = (s: string): number => {
  let letter_tracker = {};
  let count = 0;
  for (const letter of s) {
    if (Object.keys(letter_tracker).includes(letter)) {
      delete letter_tracker[letter];
      count += 2;
    } else {
      letter_tracker[letter] = 1;
    }
  }
  if (Object.keys(letter_tracker).length > 0) {
    count += 1;
  }
  return count;
};

//Optimal Solution using set()
var longestPalindromeUsingSet = function (s: string) {
  let set = new Set(),
    count = 0;
  for (let letter of s) {
    if (set.has(letter)) {
      count += 2;
      set.delete(letter);
    } else set.add(letter);
  }
  return count + (set.size > 0 ? 1 : 0);
};

File: longest_palindrome/Longest_palindrome.java
package longest_palindrome;

import java.util.HashMap;

public class Longest_palindrome {
    public int longestPalindrome(String s) {
        HashMap<Character, Integer> letter_tracker = new HashMap<>();
        int count = 0;
        char[] s_char_array = s.toCharArray();
        for (char letter : s_char_array) {
            if (letter_tracker.containsKey(letter)) {
                letter_tracker.remove(letter);
                count += 2;
            } else {
                letter_tracker.put(letter, 1);
            }
        }
        if (!letter_tracker.isEmpty()) {
            count += 1;
        }
        return count;
    }

    public int longestPalindromeOptimal(String s) {
        int[] count = new int[128];
        for (char c : s.toCharArray())
            count[c]++;

        int ans = 0;
        for (int v : count) {
            ans += v / 2 * 2;
            if (ans % 2 == 0 && v % 2 == 1)
                ans++;
        }
        return ans;
    }
}
